File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Features/VoiceSession/VoiceSessionHandle.ts
/**
 * VoiceSessionHandle.ts
 *
 * High-level voice session API for simplified voice assistant integration.
 * Handles audio capture, VAD, and processing internally.
 *
 * Matches Swift SDK: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/VoiceAgent/RunAnywhere+VoiceSession.swift
 *
 * Usage:
 * ```typescript
 * // Start a voice session
 * const session = await RunAnywhere.startVoiceSession();
 *
 * // Consume events
 * for await (const event of session.events()) {
 *   switch (event.type) {
 *     case 'listening': updateAudioMeter(event.audioLevel); break;
 *     case 'transcribed': showUserText(event.transcription); break;
 *     case 'responded': showAssistantText(event.response); break;
 *     case 'speaking': showSpeakingIndicator(); break;
 *   }
 * }
 *
 * // Or use callback
 * const session = await RunAnywhere.startVoiceSession({
 *   onEvent: (event) => { ... }
 * });
 * ```
 */

import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import { AudioCaptureManager } from './AudioCaptureManager';

// Lazy-load EventBus to avoid circular dependency issues during module initialization
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let _eventBus: any = null;
function getEventBus() {
  if (!_eventBus) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      _eventBus = require('../../Public/Events').EventBus;
    } catch {
      // EventBus not available
    }
  }
  return _eventBus;
}

/**
 * Safely publish an event to the EventBus
 * Handles cases where EventBus may not be fully initialized due to circular dependencies
 */
function safePublish(eventType: string, event: Record<string, unknown>): void {
  try {
    const eventBus = getEventBus();
    if (eventBus?.publish) {
      eventBus.publish(eventType, event);
    }
  } catch {
    // Ignore EventBus errors - events are non-critical for voice session functionality
  }
}
import { AudioPlaybackManager } from './AudioPlaybackManager';
import * as STT from '../../Public/Extensions/RunAnywhere+STT';
import * as TextGeneration from '../../Public/Extensions/RunAnywhere+TextGeneration';
import * as TTS from '../../Public/Extensions/RunAnywhere+TTS';

const logger = new SDKLogger('VoiceSession');

// ============================================================================
// Types
// ============================================================================

/**
 * Voice session configuration
 * Matches Swift: VoiceSessionConfig
 */
export interface VoiceSessionConfig {
  /** Silence duration (seconds) before processing speech (default: 1.5) */
  silenceDuration?: number;

  /** Minimum audio level to detect speech (0.0 - 1.0, default: 0.1) */
  speechThreshold?: number;

  /** Whether to auto-play TTS response (default: true) */
  autoPlayTTS?: boolean;

  /** Whether to auto-resume listening after TTS playback (default: true) */
  continuousMode?: boolean;

  /** Language code for STT (default: 'en') */
  language?: string;

  /** System prompt for LLM */
  systemPrompt?: string;

  /** Event callback (alternative to using events() iterator) */
  onEvent?: VoiceSessionEventCallback;
}

/**
 * Default configuration
 */
export const DEFAULT_VOICE_SESSION_CONFIG: Required<Omit<VoiceSessionConfig, 'onEvent'>> = {
  silenceDuration: 1.5,
  speechThreshold: 0.1,
  autoPlayTTS: true,
  continuousMode: true,
  language: 'en',
  systemPrompt: '',
};

/**
 * Voice session event types
 * Matches Swift: VoiceSessionEvent
 */
export type VoiceSessionEventType =
  | 'started'
  | 'listening'
  | 'speechStarted'
  | 'speechEnded'
  | 'processing'
  | 'transcribed'
  | 'responded'
  | 'speaking'
  | 'turnCompleted'
  | 'stopped'
  | 'error';

/**
 * Voice session event
 */
export interface VoiceSessionEvent {
  type: VoiceSessionEventType;
  timestamp: number;
  /** Audio level (for 'listening' events, 0.0 - 1.0) */
  audioLevel?: number;
  /** User's transcribed text (for 'transcribed' and 'turnCompleted' events) */
  transcription?: string;
  /** Assistant's response (for 'responded' and 'turnCompleted' events) */
  response?: string;
  /** TTS audio data (for 'turnCompleted' events) */
  audio?: string;
  /** Error message (for 'error' events) */
  error?: string;
}

/**
 * Voice session event callback
 */
export type VoiceSessionEventCallback = (event: VoiceSessionEvent) => void;

/**
 * Voice session state
 */
export type VoiceSessionState =
  | 'idle'
  | 'starting'
  | 'listening'
  | 'processing'
  | 'speaking'
  | 'stopped'
  | 'error';

// ============================================================================
// VoiceSessionHandle
// ============================================================================

/**
 * VoiceSessionHandle
 *
 * Handle to control an active voice session.
 * Manages the full voice interaction loop: listen -> transcribe -> respond -> speak.
 *
 * Matches Swift SDK: VoiceSessionHandle actor
 */
export class VoiceSessionHandle {
  private config: Required<Omit<VoiceSessionConfig, 'onEvent'>>;
  private audioCapture: AudioCaptureManager;
  private audioPlayback: AudioPlaybackManager;
  private eventCallback: VoiceSessionEventCallback | null = null;
  private eventListeners: VoiceSessionEventCallback[] = [];

  private state: VoiceSessionState = 'idle';
  private isSpeechActive = false;
  private lastSpeechTime: number | null = null;
  private vadInterval: ReturnType<typeof setInterval> | null = null;
  private currentAudioLevel = 0;

  constructor(config: VoiceSessionConfig = {}) {
    const { onEvent, ...rest } = config;
    this.config = { ...DEFAULT_VOICE_SESSION_CONFIG, ...rest };
    this.audioCapture = new AudioCaptureManager({ sampleRate: 16000 });
    this.audioPlayback = new AudioPlaybackManager();

    if (onEvent) {
      this.eventCallback = onEvent;
    }
  }

  // ============================================================================
  // Public Properties
  // ============================================================================

  /**
   * Current session state
   */
  get sessionState(): VoiceSessionState {
    return this.state;
  }

  /**
   * Whether the session is running (listening or processing)
   */
  get isRunning(): boolean {
    return this.state !== 'idle' && this.state !== 'stopped' && this.state !== 'error';
  }

  /**
   * Whether audio is currently playing
   */
  get isSpeaking(): boolean {
    return this.audioPlayback.isPlaying;
  }

  /**
   * Current audio level (0.0 - 1.0)
   */
  get audioLevel(): number {
    return this.currentAudioLevel;
  }

  // ============================================================================
  // Public Methods
  // ============================================================================

  /**
   * Start the voice session
   */
  async start(): Promise<void> {
    if (this.isRunning) {
      logger.warning('Session already running');
      return;
    }

    this.state = 'starting';
    logger.info('Starting voice session...');

    try {
      // Check if models are loaded
      const sttLoaded = await STT.isSTTModelLoaded();
      const llmLoaded = await TextGeneration.isModelLoaded();
      const ttsLoaded = await TTS.isTTSModelLoaded();

      if (!sttLoaded || !llmLoaded || !ttsLoaded) {
        throw new Error(
          `Voice agent not ready. Models loaded: STT=${sttLoaded}, LLM=${llmLoaded}, TTS=${ttsLoaded}`
        );
      }

      // Request microphone permission
      const hasPermission = await this.audioCapture.requestPermission();
      if (!hasPermission) {
        throw new Error('Microphone permission denied');
      }

      this.emit({ type: 'started', timestamp: Date.now() });
      await this.startListening();

    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      this.state = 'error';
      logger.error(`Failed to start session: ${errorMsg}`);
      this.emit({ type: 'error', timestamp: Date.now(), error: errorMsg });
      throw error;
    }
  }

  /**
   * Stop the voice session
   */
  stop(): void {
    if (this.state === 'idle' || this.state === 'stopped') {
      return;
    }

    logger.info('Stopping voice session');
    this.state = 'stopped';

    // Stop audio
    this.audioCapture.cleanup();
    this.audioPlayback.stop();

    // Clear VAD
    if (this.vadInterval) {
      clearInterval(this.vadInterval);
      this.vadInterval = null;
    }

    this.isSpeechActive = false;
    this.lastSpeechTime = null;

    this.emit({ type: 'stopped', timestamp: Date.now() });
    logger.info('Voice session stopped');
  }

  /**
   * Force process current audio (push-to-talk mode)
   */
  async sendNow(): Promise<void> {
    if (!this.isRunning) {
      logger.warning('Session not running');
      return;
    }

    this.isSpeechActive = false;
    await this.processCurrentAudio();
  }

  /**
   * Add event listener
   */
  addEventListener(callback: VoiceSessionEventCallback): () => void {
    this.eventListeners.push(callback);
    return () => {
      const index = this.eventListeners.indexOf(callback);
      if (index > -1) {
        this.eventListeners.splice(index, 1);
      }
    };
  }

  /**
   * Set single event callback (alternative to addEventListener)
   */
  setEventCallback(callback: VoiceSessionEventCallback | null): void {
    this.eventCallback = callback;
  }

  /**
   * Create async iterator for events
   * Matches Swift's AsyncStream pattern
   */
  async *events(): AsyncGenerator<VoiceSessionEvent> {
    const queue: VoiceSessionEvent[] = [];
    let resolver: ((value: VoiceSessionEvent | null) => void) | null = null;
    let done = false;

    const unsubscribe = this.addEventListener((event) => {
      if (event.type === 'stopped' || event.type === 'error') {
        done = true;
      }

      if (resolver) {
        const currentResolver = resolver;
        resolver = null;
        currentResolver(event);
      } else {
        queue.push(event);
      }
    });

    try {
      while (!done) {
        if (queue.length > 0) {
          const event = queue.shift()!;
          yield event;
          if (event.type === 'stopped' || event.type === 'error') {
            break;
          }
        } else {
          const event = await new Promise<VoiceSessionEvent | null>((resolve) => {
            resolver = resolve;
          });
          if (event === null) break;
          yield event;
        }
      }
    } finally {
      unsubscribe();
    }
  }

  /**
   * Cleanup resources
   */
  cleanup(): void {
    this.stop();
    this.audioCapture.cleanup();
    this.audioPlayback.cleanup();
    this.eventListeners = [];
    this.eventCallback = null;
    logger.info('VoiceSessionHandle cleaned up');
  }

  // ============================================================================
  // Private Methods
  // ============================================================================

  private emit(event: VoiceSessionEvent): void {
    // Call single callback
    if (this.eventCallback) {
      this.eventCallback(event);
    }

    // Call all listeners
    for (const listener of this.eventListeners) {
      try {
        listener(event);
      } catch (error) {
        logger.error(`Event listener error: ${error}`);
      }
    }

    // Publish to EventBus for app-wide observation
    // Map event type to EventBus type (note: we avoid spreading 'type' twice)
    const { type, ...eventData } = event;
    const eventBusType = `voiceSession_${type}` as const;

    switch (eventBusType) {
      case 'voiceSession_started':
        safePublish('Voice', { type: 'voiceSession_started' });
        break;
      case 'voiceSession_listening':
        safePublish('Voice', { type: 'voiceSession_listening', audioLevel: eventData.audioLevel });
        break;
      case 'voiceSession_speechStarted':
        safePublish('Voice', { type: 'voiceSession_speechStarted' });
        break;
      case 'voiceSession_speechEnded':
        safePublish('Voice', { type: 'voiceSession_speechEnded' });
        break;
      case 'voiceSession_processing':
        safePublish('Voice', { type: 'voiceSession_processing' });
        break;
      case 'voiceSession_transcribed':
        safePublish('Voice', { type: 'voiceSession_transcribed', transcription: eventData.transcription });
        break;
      case 'voiceSession_responded':
        safePublish('Voice', { type: 'voiceSession_responded', response: eventData.response });
        break;
      case 'voiceSession_speaking':
        safePublish('Voice', { type: 'voiceSession_speaking' });
        break;
      case 'voiceSession_turnCompleted':
        safePublish('Voice', {
          type: 'voiceSession_turnCompleted',
          transcription: eventData.transcription,
          response: eventData.response,
          audio: eventData.audio,
        });
        break;
      case 'voiceSession_stopped':
        safePublish('Voice', { type: 'voiceSession_stopped' });
        break;
      case 'voiceSession_error':
        safePublish('Voice', { type: 'voiceSession_error', error: eventData.error });
        break;
    }
  }

  private async startListening(): Promise<void> {
    this.state = 'listening';
    this.isSpeechActive = false;
    this.lastSpeechTime = null;

    // Set up audio level callback
    this.audioCapture.setAudioLevelCallback((level) => {
      this.currentAudioLevel = level;
      this.emit({ type: 'listening', timestamp: Date.now(), audioLevel: level });
    });

    // Start recording
    await this.audioCapture.startRecording();

    // Start VAD monitoring loop (matches Swift's startAudioLevelMonitoring)
    this.startVADMonitoring();
  }

  /**
   * VAD monitoring loop - runs every 50ms
   * Matches Swift: startAudioLevelMonitoring()
   */
  private startVADMonitoring(): void {
    this.vadInterval = setInterval(() => {
      this.checkSpeechState(this.currentAudioLevel);
    }, 50);
  }

  /**
   * Check speech state based on audio level
   * Matches Swift: checkSpeechState(level:)
   */
  private checkSpeechState(level: number): void {
    if (!this.isRunning || this.state === 'processing' || this.state === 'speaking') {
      return;
    }

    if (level > this.config.speechThreshold) {
      // Speech detected
      if (!this.isSpeechActive) {
        logger.debug('Speech started');
        this.isSpeechActive = true;
        this.emit({ type: 'speechStarted', timestamp: Date.now() });
      }
      this.lastSpeechTime = Date.now();

    } else if (this.isSpeechActive) {
      // Was speaking, now silent - check if silence is long enough
      if (this.lastSpeechTime) {
        const silenceDuration = (Date.now() - this.lastSpeechTime) / 1000;

        if (silenceDuration > this.config.silenceDuration) {
          logger.debug(`Speech ended (silence: ${silenceDuration.toFixed(2)}s)`);
          this.isSpeechActive = false;
          this.emit({ type: 'speechEnded', timestamp: Date.now() });

          // Process the audio
          this.processCurrentAudio();
        }
      }
    }
  }

  /**
   * Process current audio through the pipeline: STT -> LLM -> TTS
   */
  private async processCurrentAudio(): Promise<void> {
    // Stop VAD and recording
    if (this.vadInterval) {
      clearInterval(this.vadInterval);
      this.vadInterval = null;
    }

    this.state = 'processing';
    this.emit({ type: 'processing', timestamp: Date.now() });

    try {
      // Stop recording and get audio file
      const { path: audioPath } = await this.audioCapture.stopRecording();
      logger.info(`Audio recorded: ${audioPath}`);

      // Transcribe using STT
      const sttResult = await STT.transcribeFile(audioPath, {
        language: this.config.language,
      });
      const transcription = sttResult.text?.trim() || '';

      if (!transcription) {
        logger.info('No speech detected in audio');
        if (this.config.continuousMode && this.isRunning) {
          await this.startListening();
        }
        return;
      }

      // Emit transcription
      this.emit({
        type: 'transcribed',
        timestamp: Date.now(),
        transcription,
      });
      logger.info(`Transcribed: "${transcription}"`);

      // Generate response using LLM
      const prompt = this.config.systemPrompt
        ? `${this.config.systemPrompt}\n\nUser: ${transcription}\nAssistant:`
        : transcription;

      const llmResult = await TextGeneration.generate(prompt, {
        maxTokens: 500,
        temperature: 0.7,
      });
      const response = llmResult.text || '';

      // Emit response
      this.emit({
        type: 'responded',
        timestamp: Date.now(),
        response,
      });
      logger.info(`Response: "${response.substring(0, 100)}..."`);

      // Synthesize and play TTS if enabled
      let synthesizedAudio: string | undefined;

      if (this.config.autoPlayTTS && response) {
        this.state = 'speaking';
        this.emit({ type: 'speaking', timestamp: Date.now() });

        try {
          const ttsResult = await TTS.synthesize(response);
          synthesizedAudio = ttsResult.audioData;

          if (synthesizedAudio) {
            await this.audioPlayback.play(synthesizedAudio, ttsResult.sampleRate);
          }
        } catch (ttsError) {
          logger.warning(`TTS failed: ${ttsError}`);
        }
      }

      // Emit complete result
      this.emit({
        type: 'turnCompleted',
        timestamp: Date.now(),
        transcription,
        response,
        audio: synthesizedAudio,
      });

    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      this.state = 'error';
      logger.error(`Processing failed: ${errorMsg}`);
      this.emit({ type: 'error', timestamp: Date.now(), error: errorMsg });
      return; // Don't resume listening on error
    }

    // Resume listening if continuous mode
    if (this.config.continuousMode && this.isRunning) {
      this.state = 'listening';
      await this.startListening();
    }
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/capabilities/voice/models/voice_session_handle.dart
/// Voice Session Handle
///
/// Matches iOS VoiceSessionHandle from RunAnywhere+VoiceSession.swift
/// Provides a handle to control an active voice session with built-in audio capture
library voice_session_handle;

import 'dart:async';
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:runanywhere/capabilities/voice/models/voice_session.dart';
import 'package:runanywhere/features/stt/services/audio_capture_manager.dart';
import 'package:runanywhere/features/tts/services/audio_playback_manager.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';

/// Handle to control an active voice session
/// Matches iOS VoiceSessionHandle from RunAnywhere+VoiceSession.swift
class VoiceSessionHandle {
  final SDKLogger _logger = SDKLogger('VoiceSessionHandle');
  final VoiceSessionConfig config;

  bool _isRunning = false;
  bool _isProcessing = false;
  Uint8List _audioBuffer = Uint8List(0);
  DateTime? _lastSpeechTime;
  bool _isSpeechActive = false;

  final StreamController<VoiceSessionEvent> _eventController =
      StreamController<VoiceSessionEvent>.broadcast();

  // Audio capture manager for recording
  final AudioCaptureManager _audioCapture = AudioCaptureManager();

  // Audio playback manager for TTS output
  final AudioPlaybackManager _audioPlayback = AudioPlaybackManager();

  // Callback for processing audio (injected from RunAnywhere)
  final Future<VoiceAgentProcessResult> Function(Uint8List audioData)?
      _processAudioCallback;

  // Callback for voice agent readiness check
  final Future<bool> Function()? _isVoiceAgentReadyCallback;

  // Callback for initializing voice agent with loaded models
  final Future<void> Function()? _initializeVoiceAgentCallback;

  VoiceSessionHandle({
    VoiceSessionConfig? config,
    Future<VoiceAgentProcessResult> Function(Uint8List audioData)?
        processAudioCallback,
    @Deprecated('Permission is now handled internally by AudioCaptureManager')
    Future<bool> Function()? requestPermissionCallback,
    Future<bool> Function()? isVoiceAgentReadyCallback,
    Future<void> Function()? initializeVoiceAgentCallback,
  })  : config = config ?? VoiceSessionConfig.defaultConfig,
        _processAudioCallback = processAudioCallback,
        _isVoiceAgentReadyCallback = isVoiceAgentReadyCallback,
        _initializeVoiceAgentCallback = initializeVoiceAgentCallback;

  /// Stream of session events
  /// Matches iOS VoiceSessionHandle.events
  Stream<VoiceSessionEvent> get events => _eventController.stream;

  /// Whether the session is currently running
  bool get isRunning => _isRunning;

  /// Whether the session is currently processing audio or playing TTS
  bool get isProcessing => _isProcessing;

  /// Start the voice session
  /// Matches iOS VoiceSessionHandle.start()
  Future<void> start() async {
    if (_isRunning) {
      _logger.warning('Voice session already running');
      return;
    }

    _logger.info('🚀 Starting voice session...');

    // Check if voice agent components are ready
    _logger.info('Checking if voice agent components are ready...');
    final componentsReady = await _isVoiceAgentReadyCallback?.call() ?? false;
    _logger.info('Voice agent components ready: $componentsReady');

    if (!componentsReady) {
      const errorMsg =
          'Voice agent components not ready. Make sure STT, LLM, and TTS models are loaded.';
      _logger.error('❌ $errorMsg');
      _emit(const VoiceSessionError(message: errorMsg));
      throw const VoiceSessionException(
        VoiceSessionErrorType.notReady,
        errorMsg,
      );
    }

    // Always initialize voice agent with loaded models
    // This creates the voice agent handle and connects it to the shared component handles
    try {
      _logger.info('Initializing voice agent with loaded models...');
      await _initializeVoiceAgentCallback?.call();
      _logger.info('✅ Voice agent initialized successfully');
    } catch (e) {
      _logger.error('❌ Failed to initialize voice agent: $e');
      final errorMsg = 'Voice agent initialization failed: $e';
      _emit(VoiceSessionError(message: errorMsg));
      rethrow;
    }

    // Request mic permission via audio capture manager
    _logger.info('Requesting microphone permission...');
    final hasPermission = await _audioCapture.requestPermission();
    if (!hasPermission) {
      _logger.error('❌ Microphone permission denied');
      _emit(const VoiceSessionError(message: 'Microphone permission denied'));
      throw const VoiceSessionException(
        VoiceSessionErrorType.microphonePermissionDenied,
        'Microphone permission denied',
      );
    }
    _logger.info('✅ Microphone permission granted');

    _isRunning = true;
    _emit(const VoiceSessionStarted());

    // Start listening - this starts the audio capture loop
    await _startListening();

    _logger.info('✅ Voice session started with audio capture');
  }

  /// Start audio capture loop
  /// Matches iOS VoiceSessionHandle.startListening()
  Future<void> _startListening() async {
    if (_isProcessing) {
      _logger.warning('⚠️ Cannot start listening while processing');
      return;
    }

    _audioBuffer = Uint8List(0);
    _lastSpeechTime = null;
    _isSpeechActive = false;

    _logger.info('🎙️ Starting audio capture...');
    _logger.info(
        '📋 Config: speechThreshold=${config.speechThreshold}, silenceDuration=${config.silenceDuration}s');

    try {
      int chunkCount = 0;
      double maxLevelSeen = 0.0;
      await _audioCapture.startRecording((Uint8List audioData) {
        if (!_isRunning || _isProcessing) {
          return;
        }
        chunkCount++;

        // Log first few chunks and then periodically
        if (chunkCount <= 5 || chunkCount % 50 == 0) {
          final audioLevel = _calculateAudioLevel(audioData);
          if (audioLevel > maxLevelSeen) maxLevelSeen = audioLevel;
          _logger.info(
              '📊 Audio chunk #$chunkCount: ${audioData.length} bytes, level=${audioLevel.toStringAsFixed(4)}, max=${maxLevelSeen.toStringAsFixed(4)}, threshold=${config.speechThreshold}');
        }
        _handleAudioChunk(audioData);
      });
      _logger.info(
          '✅ Audio capture started successfully - waiting for audio data...');
    } catch (e) {
      _logger.error('❌ Failed to start audio capture: $e');
      _emit(VoiceSessionError(message: 'Failed to start audio capture: $e'));
      _isRunning = false;
      rethrow;
    }
  }

  /// Stop audio capture (used during processing/playback to prevent feedback)
  void _stopListening() {
    unawaited(_audioCapture.stopRecording());
    _audioBuffer = Uint8List(0);
    _isSpeechActive = false;
    _lastSpeechTime = null;
    _logger.info('🔇 Audio capture stopped');
  }

  /// Handle incoming audio chunk from capture
  void _handleAudioChunk(Uint8List data) {
    if (!_isRunning || _isProcessing) return;

    // Calculate audio level from the audio data
    final audioLevel = _calculateAudioLevel(data);

    // Append to buffer
    final newBuffer = Uint8List(_audioBuffer.length + data.length);
    newBuffer.setRange(0, _audioBuffer.length, _audioBuffer);
    newBuffer.setRange(_audioBuffer.length, newBuffer.length, data);
    _audioBuffer = newBuffer;

    // Check speech state with calculated audio level
    _checkSpeechState(audioLevel);
  }

  /// Calculate audio level (RMS) from audio data
  /// Returns 0.0 to 1.0
  double _calculateAudioLevel(Uint8List data) {
    if (data.isEmpty) return 0.0;

    // Audio is 16-bit PCM, so read as Int16
    final samples = data.length ~/ 2;
    if (samples == 0) return 0.0;

    double sumSquares = 0.0;
    for (int i = 0; i < samples; i++) {
      // Read little-endian Int16
      final int low = data[i * 2];
      final int high = data[i * 2 + 1];
      int sample = (high << 8) | low;
      // Handle sign extension for negative values
      if (sample > 32767) sample -= 65536;

      final normalized = sample / 32768.0;
      sumSquares += normalized * normalized;
    }

    final rms = math.sqrt(sumSquares / samples);
    // Scale to 0-1 range (RMS of full-scale sine is ~0.707)
    return math.min(1.0, rms * 1.4);
  }

  /// Stop the voice session
  /// Matches iOS VoiceSessionHandle.stop()
  void stop() {
    if (!_isRunning) return;

    _isRunning = false;
    _isProcessing = false;

    // Stop audio capture and playback
    unawaited(_audioCapture.stopRecording());
    unawaited(_audioPlayback.stop());

    _audioBuffer = Uint8List(0);
    _isSpeechActive = false;
    _lastSpeechTime = null;

    _emit(const VoiceSessionStopped());
    unawaited(_eventController.close());

    _logger.info('Voice session stopped');
  }

  /// Force process current audio (push-to-talk)
  /// Matches iOS VoiceSessionHandle.sendNow()
  Future<void> sendNow() async {
    if (!_isRunning) return;
    _isSpeechActive = false;
    await _processCurrentAudio();
  }

  /// Feed audio data to the session (for external audio sources)
  /// Can be used for custom audio capture or testing
  void feedAudio(Uint8List data, double audioLevel) {
    if (!_isRunning || _isProcessing) return;

    // Append to buffer
    final newBuffer = Uint8List(_audioBuffer.length + data.length);
    newBuffer.setRange(0, _audioBuffer.length, _audioBuffer);
    newBuffer.setRange(_audioBuffer.length, newBuffer.length, data);
    _audioBuffer = newBuffer;

    // Check speech state
    _checkSpeechState(audioLevel);
  }

  void _emit(VoiceSessionEvent event) {
    if (!_eventController.isClosed) {
      _eventController.add(event);
    }
  }

  void _checkSpeechState(double level) {
    if (_isProcessing) return;

    _emit(VoiceSessionListening(audioLevel: level));

    if (level >= config.speechThreshold) {
      if (!_isSpeechActive) {
        _logger.info(
            '🎤 Speech STARTED! level=${level.toStringAsFixed(4)} >= threshold=${config.speechThreshold}');
        _isSpeechActive = true;
        _emit(const VoiceSessionSpeechStarted());
      }
      _lastSpeechTime = DateTime.now();
    } else if (_isSpeechActive) {
      final lastTime = _lastSpeechTime;
      if (lastTime != null) {
        // Use milliseconds for accurate comparison with fractional seconds
        final silenceMs = DateTime.now().difference(lastTime).inMilliseconds;
        final thresholdMs = (config.silenceDuration * 1000).toInt();

        if (silenceMs >= thresholdMs) {
          _logger.info(
              '🔇 Speech ENDED after ${silenceMs}ms silence, buffer: ${_audioBuffer.length} bytes');
          _isSpeechActive = false;

          // Only process if we have enough audio (~0.5s at 16kHz = 16000 bytes)
          if (_audioBuffer.length > 16000) {
            _logger.info(
                '📤 Processing ${_audioBuffer.length} bytes of audio (~${(_audioBuffer.length / 32000).toStringAsFixed(1)}s)...');
            unawaited(_processCurrentAudio());
          } else {
            _logger.warning(
                '⚠️ Audio buffer too small (${_audioBuffer.length} bytes < 16000), discarding');
            _audioBuffer = Uint8List(0);
          }
        }
      }
    }
  }

  Future<void> _processCurrentAudio() async {
    final audio = _audioBuffer;
    _audioBuffer = Uint8List(0);

    if (audio.isEmpty) {
      _logger.warning('⚠️ Cannot process: audio buffer is empty');
      return;
    }

    if (!_isRunning) {
      _logger.warning('⚠️ Cannot process: session not running');
      return;
    }

    // IMPORTANT: Stop listening during processing to prevent feedback loop
    _isProcessing = true;
    _stopListening();

    final audioDuration = audio.length / 32000; // 16kHz * 2 bytes per sample
    _logger.info(
        '🔄 Processing ${audio.length} bytes (~${audioDuration.toStringAsFixed(1)}s) of audio...');
    _emit(const VoiceSessionProcessing());

    try {
      if (_processAudioCallback == null) {
        _logger.error(
            '❌ CRITICAL: No processing callback configured! This is a bug - the callback should be set when VoiceSessionHandle is created.');
        _emit(const VoiceSessionError(
            message:
                'No processing callback configured. Voice agent may not be initialized.'));
        return;
      }

      _logger.info('📞 Calling voice agent processAudio...');
      final stopwatch = Stopwatch()..start();
      final result = await _processAudioCallback!.call(audio);
      stopwatch.stop();
      _logger.info(
          '⏱️ Voice agent processing took ${stopwatch.elapsedMilliseconds}ms');

      if (!result.speechDetected) {
        _logger
            .info('🔇 No speech detected in audio (might be silence or noise)');
        // Resume listening
        if (config.continuousMode && _isRunning) {
          _logger.info('👂 Continuous mode: Resuming listening');
          _isProcessing = false;
          await _startListening();
        }
        return;
      }

      _logger.info(
          '✅ Speech detected! Transcription: "${result.transcription ?? "(empty)"}"');

      // Emit intermediate results
      if (result.transcription != null && result.transcription!.isNotEmpty) {
        _emit(VoiceSessionTranscribed(text: result.transcription!));
      } else {
        _logger.warning('⚠️ STT returned empty transcription');
      }

      if (result.response != null && result.response!.isNotEmpty) {
        final previewLen =
            result.response!.length > 100 ? 100 : result.response!.length;
        _logger.info(
            '💬 LLM Response (${result.response!.length} chars): "${result.response!.substring(0, previewLen)}${result.response!.length > 100 ? "..." : ""}"');
        _emit(VoiceSessionResponded(text: result.response!));
      } else {
        _logger.warning('⚠️ LLM returned empty response');
      }

      // Play TTS audio if available and enabled
      if (config.autoPlayTTS &&
          result.synthesizedAudio != null &&
          result.synthesizedAudio!.isNotEmpty) {
        // TTS audio from ONNX Piper is typically 22050Hz mono PCM16
        final ttsDuration = result.synthesizedAudio!.length / (22050 * 2);
        _logger.info(
            '🔊 Playing TTS audio: ${result.synthesizedAudio!.length} bytes (~${ttsDuration.toStringAsFixed(1)}s)');
        _emit(const VoiceSessionSpeaking());

        try {
          // Play audio and wait for completion
          await _audioPlayback.play(
            result.synthesizedAudio!,
            sampleRate: 22050, // ONNX Piper TTS default
            numChannels: 1,
          );
          _logger.info('🔊 TTS playback completed');
        } catch (e) {
          _logger.error('❌ TTS playback failed: $e');
          // Continue even if playback fails
        }
      }

      // Emit complete result
      _emit(VoiceSessionTurnCompleted(
        transcript: result.transcription ?? '',
        response: result.response ?? '',
        audio: result.synthesizedAudio,
      ));
      _logger.info('✅ Voice turn completed successfully');
    } catch (e, stack) {
      _logger.error('❌ Processing failed: $e');
      _logger.error('Stack trace: $stack');
      _emit(VoiceSessionError(message: e.toString()));
    } finally {
      // Resume listening if continuous mode and session still running
      _isProcessing = false;
      if (config.continuousMode && _isRunning) {
        _logger.info('👂 Continuous mode: Resuming listening after turn');
        try {
          await _startListening();
        } catch (e) {
          _logger.error('❌ Failed to resume listening: $e');
        }
      }
    }
  }

  /// Dispose resources
  Future<void> dispose() async {
    stop();
    await _audioPlayback.dispose();
    _audioCapture.dispose();
  }
}

/// Result from voice agent processing
class VoiceAgentProcessResult {
  final bool speechDetected;
  final String? transcription;
  final String? response;
  final Uint8List? synthesizedAudio;

  const VoiceAgentProcessResult({
    required this.speechDetected,
    this.transcription,
    this.response,
    this.synthesizedAudio,
  });
}
